pkg/devfile/convert/convert.go

```diff
@@ -278,6 +330,57 @@ func setDevfileComponentsForS2I(d data.DevfileData, s2iImage string, localConfig
 		volumeMounts = append(volumeMounts, volumeMount)
 	}
 
+	// Add volumes for:
+	// - /opt/app-root
+	// - ${ODO_S2I_DEPLOYMENT_DIR} if outside of /opt/app-root
+
+	var deploymentDir string
+	for _, env := range s2iEnv {
+		if env.Name == "ODO_S2I_DEPLOYMENT_DIR" {
+			deploymentDir = env.Value
+			break
+		}
+	}
+
+	separateDeploymentsMount := len(deploymentDir) > 0 && !strings.HasPrefix(deploymentDir, occlient.DefaultAppRootDir)
+
+	volumeAppRoot := devfilev1.Component{
+		Name: "app-root-volume",
+		ComponentUnion: devfilev1.ComponentUnion{
+			Volume: &devfilev1.VolumeComponent{
+				Volume: devfilev1.Volume{
+					Size: "1Gi",
+				},
+			},
+		},
+	}
```

Comment:
I might be nitpicking, but do you think we can move the following piece of code into a local function that accepts the `Name`? I see the same code is used if `separateDeploymentsMount` is true.
```go
devfilev1.Component{
		Name: "app-root-volume",
		ComponentUnion: devfilev1.ComponentUnion{
			Volume: &devfilev1.VolumeComponent{
				Volume: devfilev1.Volume{
					Size: "1Gi",
				},
			},
		},
	}
```

Reply:
It would be nice to have some high-level functions on top of the devfile library for these creations of ComponentUnion, but it would be better to make the changes in the whole sources